File: Gstreamer-Applications/gst_streaming/tcp_server.c
/*
* Copyright (c) 2020 Qualcomm Innovation Center, Inc.  All Rights Reserved.
*
* SPDX-License-Identifier: BSD-3-Clause-Clear
*/

/*
 * tcp_server.c:
 *
 * A sample app based on gstreamer
 * The purpose is helping users to learn how to implement the
 * tcp streaming server of ISP camera with gstreamer on the
 * Qualcomm platform through this sample app.
 * 
 * Client should use VLC player to play the url:
 *    tcp://<ip>:<port>/
 */

#include <gst/gst.h>
#include <glib.h>


static void usage(const char* argv0)
{
    g_print("Usage: %s [camera id: 0|1|2|3] IP port\n", argv0);
    g_print(" IP: [x.x.x.x] : IP address of the board. It maybe through eht0 or wlan0.\n");
    g_print("Example:\n");
    g_print("    %s 0 192.168.3.5 5000\n", argv0);
}

static gboolean bus_callback(GstBus *bus, GstMessage *msg, gpointer data)
{
    GMainLoop *loop = (GMainLoop *)data;

    int type = GST_MESSAGE_TYPE(msg);
    
    if (type == GST_MESSAGE_ERROR) {
        gchar  *debug;
        GError *error;
        gst_message_parse_error(msg, &error, &debug);
        g_free(debug);
        g_printerr("Error: %s\n", error->message);
        g_error_free(error);
        g_main_loop_quit(loop);
    }

    return TRUE;
}



int main(int argc, char *argv[])
{
    GMainLoop *loop;
    GstElement *pipeline, *source, *framefilter, *parse, *muxer, *queue, *sink;
    GstBus *bus;
    guint bus_id;
    gint8 camera_id;
    gint port;

    gst_init(&argc, &argv);

    loop = g_main_loop_new(NULL, FALSE);

    if (argc != 4) {
        usage(argv[0]);
        return -1;
    }

    camera_id = atoi(argv[1]);
    if (camera_id < 0 || camera_id > 3) {
        g_printerr("camera id incorrect.\n");
        usage(argv[0]);
        return -1;
    }

    port = atoi(argv[3]);
    if (port < 0) {
        g_printerr("port incorrect.\n");
        usage(argv[0]);
        return -1;
    }

    pipeline        = gst_pipeline_new("video-streaming");
    source          = gst_element_factory_make("qtiqmmfsrc",   "qmmf-source");
    framefilter     = gst_element_factory_make("capsfilter",   "frame-filter");
    parse           = gst_element_factory_make("h264parse",    "h264-parse");
    muxer           = gst_element_factory_make("mpegtsmux",    "mpegts-muxer");
    queue           = gst_element_factory_make("queue",        "queue");
    sink            = gst_element_factory_make("tcpserversink","tcp-server");


    if (!pipeline || !source || !framefilter || !parse || !queue || !sink) {
        g_printerr("Create element failed.\n");
        return -1;
    }

    g_object_set (G_OBJECT(source), "camera", camera_id, NULL);

    if (camera_id == 0) {
        g_object_set(G_OBJECT(framefilter), "caps", 
                            gst_caps_from_string("video/x-h264,framerate=30/1,width=1920,height=1080"), NULL);
    } else {
        g_object_set(G_OBJECT(framefilter), "caps", 
                            gst_caps_from_string("video/x-h264,framerate=30/1,width=1280,height=720"), NULL);
    }

    g_object_set (G_OBJECT(parse), "config-interval", 1, NULL);
    g_object_set (G_OBJECT(sink), "host", argv[2], "port", port, NULL);


    bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
    bus_id = gst_bus_add_watch(bus, bus_callback, loop);
    gst_object_unref(bus);

    gst_bin_add_many(GST_BIN(pipeline), source, framefilter, parse, muxer, queue, sink, NULL);
    gst_element_link_many(source, framefilter, parse, muxer, queue, sink, NULL);

    gst_element_set_state(pipeline, GST_STATE_PLAYING);

    g_print("Start\n");
    g_print("  vlc tcp://%s:%d/\n", argv[2], port);
    g_main_loop_run(loop);
    g_print("Stop\n");
    gst_element_set_state(pipeline, GST_STATE_NULL);

    gst_object_unref(GST_OBJECT (pipeline));
    g_source_remove(bus_id);
    g_main_loop_unref(loop);

    gst_deinit();

    return 0;
}

File: gst-python-samples/Gstreamer-Applications/gst_streaming/tcp_server.py
import sys
import gi

gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")

from gi.repository import Gst, GLib

if __name__ == "__main__":
    # Initialize GStreamer passing command line argument
    Gst.init(sys.argv)
    loop = GLib.MainLoop()

if len(sys.argv) != 4:
    print("Missing <output-location> parameter")
    sys.exit(1)
else:
    camera_id = int(sys.argv[1])
    if (camera_id < 0 or camera_id > 3):
        print("camera id incorrect")
        sys.exit(1)
    port = int(sys.argv[3])
    if (port < 0):
        print("port incorrect.")
        sys.exit(1)

# Create the empty pipeline
pipeline = Gst.Pipeline.new("video-streaming")

# Create the elements
source = Gst.ElementFactory.make("qtiqmmfsrc", "qmmf-source")
framefilter = Gst.ElementFactory.make("capsfilter",   "frame-filter")
parse = Gst.ElementFactory.make("h264parse",    "h264-parse")
muxer = Gst.ElementFactory.make("mpegtsmux",    "mpegts-muxer")
queue = Gst.ElementFactory.make("queue",        "queue")
sink = Gst.ElementFactory.make("tcpserversink",  "tcp-server")

if not pipeline or not source or not framefilter or not parse or not queue or not sink:
    print("Create element failed.")
    sys.exit(1)

# set properties to the elements
source.set_property("camera", camera_id)

if camera_id == 0:
    framefilter.set_property("caps", Gst.Caps.from_string(
        "video/x-h264, framerate=30/1,width=1920, height=1080"))
else:
    framefilter.set_property("caps", Gst.Caps.from_string(
        "video/x-h264, framerate=30/1,width=1280, height=720"))

parse.set_property("config-interval", 1)
sink.set_property("host", sys.argv[2])
sink.set_property("port", port)

# Build the pipeline
pipeline.add(source)
pipeline.add(framefilter)
pipeline.add(parse)
pipeline.add(muxer)
pipeline.add(queue)
pipeline.add(sink)

if not Gst.Element.link(source, framefilter):
    print("ERROR: Could not link source to framefilter")
    sys.exit(1)
if not Gst.Element.link(framefilter, parse):
    print("ERROR: Could not link framefilter to parse")
    sys.exit(1)
if not Gst.Element.link(parse, muxer):
    print("ERROR: Could not link parse to muxer")
    sys.exit(1)
if not Gst.Element.link(muxer, queue):
    print("ERROR: Could not link muxer to queue")
    sys.exit(1)
if not Gst.Element.link(queue, sink):
    print("ERROR: Could not link queue to sink")
    sys.exit(1)

try:
    pipeline.set_state(Gst.State.PLAYING)
    print("Streaming started...")
    print("vlc tcp://%s:%d/" % (sys.argv[2], port))
    print('Press Ctrl+C to stop')
    loop.run()
except KeyboardInterrupt:
    print("Streaming stopped due to keyboard Interrupt")
    pipeline.set_state(Gst.State.NULL)
    sys.exit(0)
